SwiftCompilerSources/Sources/SIL/ForwardingInstruction.swift

```diff
@@ -0,0 +1,155 @@
+//===--- ForwardingInstruction.swift - forwarding instruction protocols ---===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SILBridging
+
+/// An instruction that forwards ownership from operands to results.
+public protocol ForwardingInstruction : Instruction {}
+
+extension ForwardingInstruction {
+  public var singleForwardedOperand: Operand? {
+    switch self {
+    case is StructInst, is TupleInst, is LinearFunctionInst,
+      is DifferentiableFunctionInst:
+      return nil
+    case let mdi as MarkDependenceInst:
+      return mdi.valueOperand
+    case let rtboi as RefToBridgeObjectInst:
+      return rtboi.convertedOperand
+    case let tpei as TuplePackExtractInst:
+      return tpei.tupleOperand
+    case let sei as SelectEnumInst:
+      return sei.enumOperand
+    default:
+      let realOperands = self.realOperands
+      if realOperands.isEmpty {
+        // This can happen with enum instructions that have no payload.
+        return nil
+      }
+      assert(realOperands.count == 1);
+      return realOperands[0];
+    }
+  }
+
+  public var forwardedOperands: OperandArray {
+    // Some instructions have multiple real operands but only forward one.
+    if let singleForwardingOp = singleForwardedOperand {
+      return OperandArray(base: singleForwardingOp, count: 1)
+    }
+    // All others forward all operands (for enum, this may be zero operands).
+    return operands
+  }
+
+  /// If forwarding ownership is owned, then the instruction moves an owned operand to its result, ending its lifetime. If forwarding ownership is guaranteed, then the instruction propagates the lifetime of its borrows operand through its result.
+  ///
+  /// The resulting forwarded value's ownership (Value.ownership) is not identical to the instruction's forwarding ownership property. It differs when the result is trivial type. e.g. an owned or guaranteed value can be cast to a trivial type using owned or guaranteed forwarding.
+  public var forwardingOwnership: Ownership {
+    Ownership(bridged: bridged.ForwardingInst_forwardingOwnership())
```

Comment:
Are you going to implement these properties without bridging, too?

Reply:
If I knew how to do that I would. Implementing all these bridging hook was incredibly annoying.

Reply:
I implemented everything in Swift that didn't require reading a C++ field from the instruction class. I did not bridge `isAddressOnly` yet because we need to bridge it from SILType first.